### PR TITLE
[PyTorch] Add Borrowed<Tensor{,Base}>

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95418

Yet another ownership-motivated smart pointer. This one is Tensor-specific, at least for now, and should allow us to manufacture `const Tensor&` from `at::TensorImpl*`.

Differential Revision: [D43544927](https://our.internmc.facebook.com/intern/diff/D43544927/)

### CODE REVIEW

File: c10/util/Borrowed.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  Borrowed& operator=(const T& rhs) {
+    *this = Borrowed(rhs);
+    return *this;
+  }
+
+  explicit operator bool() const noexcept {
+    return !BT::isNullRepr(repr_);
+  }
+
+  const T& operator*() const {
```
Comment (swolchok) [inline c10/util/Borrowed.h:70]: probably going to need to come back and also add `operator const T&` as @ezyang outlined way back in #14442 , but I'm happy to do that when it's needed

**Comment by swolchok** [comment]
Some patterns we should see if clang-tidy catches (a pointer to how to check that would be great):

```
void someFunc(const Tensor&);
void assignToTensorReference(const Tensor& aTensor) {
  const Tensor& t = Borrowed<Tensor>(aTensor); // dangles!
  someFunc(t);
}
```

```
const Tensor& sillyIdentity(const Tensor& t) {
  return Borrowed<Tensor>(t);
}
```

**Comment by swolchok** [comment]
we absolutely *need* to catch those patterns because we have code that currently does this:

```
  c10::MaybeOwned<Tensor> weight_maybe_owned = at::borrow_from_optional_tensor(weight_opt);
  const Tensor& weight = *weight_maybe_owned;
```

these will all dangle if we migrate to pointer-tagging MaybeOwned where `operator*` returns `Borrowed<Tensor>`.

**Comment by swolchok** [comment]
we absolutely *need* to catch those patterns because we have code that currently does this:

```
  c10::MaybeOwned<Tensor> weight_maybe_owned = at::borrow_from_optional_tensor(weight_opt);
  const Tensor& weight = *weight_maybe_owned;
```

these will all dangle if we migrate to pointer-tagging MaybeOwned where `operator*` returns `Borrowed<Tensor>`.

**Comment by ezyang** [comment]
OK, so bad news on clang-tidy: it's not actually being run on all files in the project. This is not fatal, but we will need some way to either turn it on everywhere, or enforce that you've turned it on any file that is being used here.

**Comment by ezyang** [comment]
clang-tidy under the current configuration doesn't catch the dangling bug. I tested by patching in:

```
diff --git a/.lintrunner.toml b/.lintrunner.toml
index dd94aae4a1d..1a819b55491 100644
--- a/.lintrunner.toml
+++ b/.lintrunner.toml
@@ -223,6 +223,7 @@ include_patterns = [
     'torch/csrc/onnx/**/*.cpp',
     'torch/csrc/tensor/**/*.cpp',
     'torch/csrc/utils/**/*.cpp',
+    'aten/src/ATen/test/Borrowed_test.cpp',
 ]
 exclude_patterns = [
     # The negative filters below are to exclude files that include onnx_pb.h or
diff --git a/aten/src/ATen/test/Borrowed_test.cpp b/aten/src/ATen/test/Borrowed_test.cpp
index eb85b5e99f8..3197ea86380 100644
--- a/aten/src/ATen/test/Borrowed_test.cpp
+++ b/aten/src/ATen/test/Borrowed_test.cpp
@@ -135,4 +135,8 @@ TYPED_TEST(BorrowedTest, TensorAssignment) {
   }
 }
 
+void f(const at::Tensor& t) {
+  const at::Tensor& x = *c10::Borrowed<at::Tensor>(t);
+}
+
 } // namespace
```

and running `lintrunner --take CLANGTIDY` (you need to have built PyTorch enough so all the generated files exist, and our clang-tidy config doesn't work on OSX). The enabled checks are in `.clang-tidy`

**Comment by ezyang** [comment]
I think we need https://clang.llvm.org/extra/clang-tidy/checks/bugprone/dangling-handle.html with a custom class list that includes our handle classes.

File: c10/util/Borrowed.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    return BT::referenceFromRepr(repr_);
+  };
+
+  const T* operator->() const {
+    return get();
+  }
+
+  const T* get() const {
+    return BT::getImpl(repr_);
```
Comment (peterbell10) [inline c10/util/Borrowed.h:79]: `pointerFromRepr` to complete the analogy with `MaybeOwnedTraits`?

File: c10/util/BorrowedTensorTraits.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+template <typename TensorType>
+struct BorrowedTraits {
+  using repr_type = TensorType;
+  using raw_impl_pointer_type = TensorImpl*;
+
+  static TensorType nullRepr() {
+    // intrusive_ptr(std::nullptr_t) sets the internal representation
+    // to NullType::singleton() instead. We want actual nullptr bit
+    // pattern. We will not dereference it or run the
+    // TensorImpl/intrusive_ptr destructor.
```
Comment (peterbell10) [inline c10/util/BorrowedTensorTraits.h:21]: Mixing two different null representations seems fairly dangerous. Why do we need `nullptr` here instead of `NullType::singleton()`?

Comment (swolchok) [reply c10/util/BorrowedTensorTraits.h:21]: to be able to differentiate "unset borrowed" from "set to undefined tensor"

Comment (swolchok) [reply c10/util/BorrowedTensorTraits.h:21]: this is needed in order to implement `operator bool`. Also, it's not dangerous -- the nullptr bit pattern is never to be used as a Tensor; it indicates the lack of a Tensor.

**Comment by swolchok** [comment]
> we absolutely need to catch those patterns

technically, we don't *need* it unless we go on to use Borrowed in MaybeOwned. That was my motivation for writing Borrowed, but it does solve the "manufacture a const Tensor&" problem by itself.

File: c10/util/BorrowedTensorTraits.h (hunk 3 above)
Comment (peterbell10) [reply c10/util/BorrowedTensorTraits.h:21]: > this is needed in order to implement `operator bool`

Let me rephrase. Why is it desirable to have optional-like semantics, instead of defaulting to undefined?

> Also, it's not dangerous -- the nullptr bit pattern is never to be used as a Tensor; 

That's what makes it dangerous. The `nullpt`-valued `Tensor` can escape if used incorrectly and the `intrusive_ptr` code that explicity checks for `NullType::singleton` instead of `nullptr` will interpret it as a valid object and try to `decref` it.

**Review by ezyang — APPROVED** [review]
I'm OK with giving this a shot.

File: c10/util/BorrowedTensorTraits.h (hunk 3 above)
Comment (swolchok) [reply c10/util/BorrowedTensorTraits.h:21]: > The nullptr-valued Tensor can escape if used incorrectly and the intrusive_ptr code that explicity checks for NullType::singleton instead of nullptr will interpret it as a valid object and try to decref it.

Yes, if people use this primitive incorrectly then they will get crashes.

> . Why is it desirable to have optional-like semantics, instead of defaulting to undefined?

TensorRef in functorch needs it, see https://github.com/pytorch/pytorch/pull/95419

Comment (peterbell10) [reply c10/util/BorrowedTensorTraits.h:21]: > TensorRef in functorch needs it, see https://github.com/pytorch/pytorch/pull/95419

Are you saying that `TensorRef` happens to make this distinction, or it's important to the functionality? i.e. if you instead wrote `using TensorRef = at::OptionalTensorRef` would the tests fail?
